### PR TITLE
[M] Refactor to use central Redux Store (#128)

### PR DESCRIPTION
**Pull request by burhan-syed**
In an effort to centralize state related to the API and reduce overall API calls I've introduced a central redux store with redux toolkit. 

The store and slices of state lives in `/client/src/state`
App.jsx shows some example of how to use the dispatchers to fetch data from the API and save it to the store. 
Then the state in store can be received with selectors, as shown in App.jsx or also in my root component at `client/src/components/ProductDetails/index.jsx`.

### CODE REVIEW

File: client/src/components/ProductDetails/index.jsx [hunk 1]
```diff
@@ -8,56 +10,33 @@ import ReviewStars from '../common/ReviewStars';
...
+  const productInfo = useSelector(
+    (state) => state.productInfo.data?.[productId]
+  );
+  const productStyles = useSelector(
+    (state) => state.productStyles.data?.[productId]
+  );
+  const reviewMeta = useSelector((state) => state.reviewMeta.data?.[productId]); 
+  const numReviews = reviewMeta?.ratingsSummary?.count; 
+  const reviewRating = reviewMeta?.ratingsSummary?.average;
+
```
Comment (justcord10) [inline client/src/components/ProductDetails/index.jsx:27]: I see that in Redux states are replaced by useSelectors which i assume gets the state from the central store

File: client/src/index.jsx [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
-root.setAttribute('id', 'root');
+const root = document.createElement("div");
+root.setAttribute("id", "root");
 document.body.appendChild(root);
 
-render(<App />, root);
+render(
+  <Provider store={store}>
+    <App />
+  </Provider>,
```
Comment (justcord10) [inline client/src/index.jsx:14]: oh interesting so provider is just a native Redux component to provide the store data?

File: client/src/components/ProductDetails/index.jsx [hunk 3]
```diff
@@ -8,56 +10,33 @@ import ReviewStars from '../common/ReviewStars';
...
-        })
-        .catch((err) => {
-          console.error('product detail reviews err?', err);
-        });
+    if (productStyles?.results) {
+      setSelectedStyle(() => {
+        let defaultStyle = productStyles.results?.find((style) => style?.['default?'] === true);
+        return defaultStyle || productStyles.results[0];
+      });
     }
```
Comment (justcord10) [inline client/src/components/ProductDetails/index.jsx:34]: also see that all axios calls are offloaded to the store and only basic logic should remain

File: client/src/state/products/productInfoSlice.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      state.isLoading = true
+    });
+    builder.addCase(fetchProductInfo.fulfilled, (state, action) => {
+      state.isLoading = false
+      state.data = {...state.data, [action.payload.id]: action.payload }
+    });
+    builder.addCase(fetchProductInfo.rejected, (state, action) => {
+      state.isLoading = false
+      state.error = action.error.message
+    });
```
Comment (justcord10) [inline client/src/state/products/productInfoSlice.js:33]: this file i would need an explanation but it looks like this is just the way to add the data from the axios call to the store?

**Review by justcord10 — APPROVED** [review]
Redux seems fairly simple to implement. but looks good, clean code as always just need a quick confirmation on how the slicers work.

File: client/src/state/products/productInfoSlice.js (hunk 4 above)
Comment (burhan-syed) [reply client/src/state/products/productInfoSlice.js:33]: Yes, when fetchProductInfo is dispatched redux toolkit uses the extra reducers defined here to handle different states of the dispatched event created by createAsyncThunk. So when it is fulfilled the defined callback on line 26 is called and redux's state updated with the data returned from the function defined in fetchProductInfo (attached to the action.payload).

**Pull request merged**